x-pack/plugins/security/server/lib/authentication/providers/saml.ts

```diff
@@ -151,9 +136,20 @@ export class SAMLAuthenticationProvider {
       }
     }
 
-    if (authenticationResult.notHandled()) {
-      // Let's check if user is redirected to Kibana from IdP with valid SAMLResponse.
-      authenticationResult = await this.authenticateViaPayload(request, state);
+    // Let's check if user is redirected to Kibana from IdP with valid SAMLResponse.
+    if (isSAMLResponsePayload(request.payload)) {
+      if (authenticationResult.notHandled()) {
+        authenticationResult = await this.authenticateViaPayload(request, state);
+      } else if (authenticationResult.succeeded()) {
+        // If user has been authenticated via session, but request also includes SAML payload
+        // we should check whether this payload is for the exactly same user and if not
+        // we'll re-authenticate user and forward to a page with the respective warning.
+        authenticationResult = await this.authenticateViaNewPayload(
```

Comment:
__note:__ couldn't find a better name for `authenticateViaNewPayload` :/

Reply:
When reading through the code and seeing this name and the comment, I wasn't surprised, so it seems like a find name for the time being.